### ISSUE TITLE
Errors are hidden from observabiilty tooling

### ISSUE THREAD

**Issue by jedateach** [CrowdStrike/ember-headless-form#529]
If I provide an action like:
```js
@action async onSubmit() {
  throw new Error("this doesn't bubble up")
}
````
I would expect the failure to bubble out of the ember app to my observability/logging tools.
However, this isn't the case.

I'm not sure if this is by design.

Ember headless forms wraps the `[validate](https://github.com/CrowdStrike/ember-headless-form/blob/main/packages/ember-headless-form/src/components/headless-form.gts#L357-L359)` and `[onSubmit](https://github.com/CrowdStrike/ember-headless-form/blob/main/packages/ember-headless-form/src/components/headless-form.gts#L470-L472)` callbacks in 

This approach makes sense for being able to present expected errors in templates.
But what about unexpected failures / bugs that you want to be seen in devTools / observability tools?

[ember-async-data readme comes](https://github.com/tracked-tools/ember-async-data?tab=readme-ov-file#unit-testing) with this statement:

> Handling errors is slightly more complicated: TrackedAsyncData “re-throws” the promises it works with when they have errors, to avoid silently swallowing them in a way that prevents you from using them with logging infrastructure or otherwise dealing with them in your app’s infrastructure. However, this means you must also account for them in your testing

This test verifies the submission failure state works:
https://github.com/CrowdStrike/ember-headless-form/blob/main/test-app/tests/integration/components/headless-form-async-test.gts#L222
...but not any bubbling out.

I'm guessing we want to have these errors bubble out.

**Comment by simonihmig** [comment]
First, this behavior is not by design. As you can see in the place that you [linked to](https://github.com/CrowdStrike/ember-headless-form/blob/main/packages/ember-headless-form/src/components/headless-form.gts#L470-L472) where we call onSubmit, we are not having any catch in place that would hide the rejected promise.

You are quoting ember-async-data that it “re-throws the promises", but I actually don't see that in their code. It does catch an error [here](https://github.com/tracked-tools/ember-async-data/blob/main/ember-async-data/src/tracked-async-data.ts#L57-L60), but without rethrowing. So I am not really sure what they refer to here. I _think_ if they would rethrow, that would also change that behavior, as again we don't do any catch on our side.

What would be your expectation for the rejected errors to not be hidden? That you see them triggering a `unhandledrejection` event on `window`?

**Comment by jedateach** [comment]
Yes, that was my expectation when developing with ember-headless-form. I was struggling to debug an issue, because I had no feedback.
My use case is to perform an action beyond the form, so I hadn't wired up the isRejected check to show in the UI.

I suppose there could be valid reasons to reject the promise and present this in the form, but equally there might be unexpected bugs where that isn't suitable.

Would it be worth having a specific type of Error to throw in each circumstance? Catch all instances of FormError for presenting in ui, but throw anything else?
Probably a breaking change tough.

**Comment by sukima** [comment]
The confusion comes from the difference between errors as flow control versus exceptions should be exceptional.

For example, the current implementation assumes any promise rejection in the onSubmit handler is meant for flow control (render the error in the template). However, typically exceptions in general are meant more to crash an app. If there was no expected errors and the use of `submissionState.isRejected` is not used there is no way for the runtime to know about the unexpected error. And thus gets swallowed. This can be unfortunate as a larger call stack for submission could easily `throw Error` and that would be a accepted best practice not realizing that the error turns from unexpected to expected at the submission call-site.

What this means is that in code that handles its _expected_ errors but then throws its _unexpected_ errors assuming it will bubble to the `onerror`/unhanded promise rejects like DOM event handlers typically do they go silent (swallowed).

A warning explaining that every use of `@onSubmit` _must_ involve rendering `submissionState.error` or any exceptions will be lost is needed in the documentation. Better yet, offer a way to escape-hatch or opt-in to this rejection state. Perhaps allow the return value to be something more expressive, or pass in a `submissionError()` callback that can render to the template while a promise rejection does its normal crash the system / log to console.

As it is now I had to get bit by missing console logs before I found this issue that explains what was happening.
